_extensions/custom_roles.py

```diff
@@ -19,12 +25,16 @@ def setup(app):
 
 def autolink(pattern):
     def role(name, rawtext, text, lineno, inliner, options={}, content=[]):
+        """Combine literal + reference (unless the text is prefixed by a !)."""
         if " " in text:
             url_text = urllib.parse.quote(f"{text}")
         else:
             url_text = text
         url = pattern % (url_text,)
-        node = nodes.reference(rawtext, text, refuri=url, **options)
+        node = nodes.literal('', text.lstrip('!'), **options)
```

Comment:
```suggestion
        node = nodes.literal(rawtext, text.lstrip('!'), **options)
```

The source text should be included here, as it was before—IIRC, it is used at least in error messages, and possibly by translations and other extensions

Reply:
Good catch!